File: .tmp_sc4013/java_22/3933_largest_local_values_in_a_matrix_ii.java
// CONFIG class=Solution method=countLocalMaximums types=None
// LeetCode 3933 - Largest Local Values in a Matrix II
// https://leetcode.com/problems/largest-local-values-in-a-matrix-ii/

import java.util.ArrayList;
import java.util.List;

class Solution {
    public int countLocalMaximums(int[][] matrix) {
        int rows = matrix.length, cols = matrix[0].length;
        List<int[]>[] positions = new ArrayList[201];
        for (int i = 0; i < 201; i++) positions[i] = new ArrayList<>();
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                int value = matrix[row][col];
                if (value > 0) positions[value].add(new int[] { row, col });
            }
        }
        int answer = 0;
        for (int value = 1; value <= 200; value++) {
            if (positions[value].isEmpty()) continue;
            int[][] prefix = new int[rows + 1][cols + 1];
            for (int row = 0; row < rows; row++) {
                for (int col = 0; col < cols; col++) {
                    int add = matrix[row][col] > value ? 1 : 0;
                    prefix[row + 1][col + 1] = prefix[row][col + 1] + prefix[row + 1][col] - prefix[row][col] + add;
                }
            }
            for (int[] pos : positions[value]) {
                int row = pos[0], col = pos[1];
                int top = Math.max(0, row - value), bottom = Math.min(rows - 1, row + value);
                int left = Math.max(0, col - value), right = Math.min(cols - 1, col + value);
                int greater = prefix[bottom + 1][right + 1] - prefix[top][right + 1] - prefix[bottom + 1][left] + prefix[top][left];
                for (int dr : new int[] { -value, value }) {
                    for (int dc : new int[] { -value, value }) {
                        int rr = row + dr, cc = col + dc;
                        if (rr >= 0 && rr < rows && cc >= 0 && cc < cols && matrix[rr][cc] > value) greater--;
                    }
                }
                if (greater == 0) answer++;
            }
        }
        return answer;
    }
}

File: .tmp_py4013/batch_17_dump/3933_largest_local_values_in_a_matrix_ii.js
// LeetCode 3933 - Largest Local Values in a Matrix II
// https://leetcode.com/problems/largest-local-values-in-a-matrix-ii/
var countLocalMaximums = function(matrix) {
        let rows = matrix.length, cols = matrix[0].length;
        let positions = Array.from({length: 201}, () => []);
        for (let i = 0; i < 201; i++) positions[i] = [];
        for (let row = 0; row < rows; row++) {
            for (let col = 0; col < cols; col++) {
                let value = matrix[row][col];
                if (value > 0) positions[value].push([ row, col ]);
            }
        }
        let answer = 0;
        for (let value = 1; value <= 200; value++) {
            if (positions[value].length === 0) continue;
            let prefix = new Array(rows + 1).fill(0)[cols + 1];
            for (let row = 0; row < rows; row++) {
                for (let col = 0; col < cols; col++) {
                    let add = matrix[row][col] > value ? 1 : 0;
                    prefix[row + 1][col + 1] = prefix[row][col + 1] + prefix[row + 1][col] - prefix[row][col] + add;
                }
            }
            for (const pos of positions[value]) {
                let row = pos[0], col = pos[1];
                let top = Math.max(0, row - value), bottom = Math.min(rows - 1, row + value);
                let left = Math.max(0, col - value), right = Math.min(cols - 1, col + value);
                let greater = prefix[bottom + 1][right + 1] - prefix[top][right + 1] - prefix[bottom + 1][left] + prefix[top][left];
                for (const dr of [ -value, value ]) {
                    for (const dc of [ -value, value ]) {
                        let rr = row + dr, cc = col + dc;
                        if (rr >= 0 && rr < rows && cc >= 0 && cc < cols && matrix[rr][cc] > value) greater--;
                    }
                }
                if (greater == 0) answer++;
            }
        }
        return answer;
    
};
